File: packages/desktop-helpers/windows/src/ipc/router.cpp
// router.cpp

#include "router.hpp"

#include <exception>
#include <string>

#include "event_emitter.hpp"
#include "../control/control_handlers.hpp"
#include "../screen/screen_handlers.hpp"
#include "../ax/ax_handlers.hpp"
#include "../ax/ax_subscribe_handlers.hpp"
#include "../foreground/foreground_handlers.hpp"
#include "../ocr/ocr_handlers.hpp"
#include "../audio/audio_handlers.hpp"
#include "../permissions/permission_handlers.hpp"
#include "../util/logger.hpp"

namespace corivo::ipc {

namespace {

void send_error(const std::string& id, const char* code, const std::string& message) {
    corivo::proto::Response resp;
    resp.id = id;
    resp.ok = false;
    corivo::proto::ResponseError err;
    err.code = code;
    err.message = message;
    resp.error = err;
    try {
        EventEmitter::instance().send(resp);
    } catch (const std::exception& e) {
        log::error(std::string("router: response write failed: ") + e.what());
    }
}

} // namespace

void Router::dispatch(const corivo::proto::Request& request) {
    try {
        if (request.method == corivo::proto::methods::PING) {
            corivo::control::ping(request);
        } else if (request.method == corivo::proto::methods::SHUTDOWN) {
            corivo::control::shutdown(request);
        } else if (request.method == corivo::proto::methods::SCREEN_CAPTURE) {
            corivo::screen::handle_capture(request);
        } else if (request.method == corivo::proto::methods::SCREEN_LIST_DISPLAYS) {
            corivo::screen::handle_list_displays(request);
        } else if (request.method == corivo::proto::methods::AX_QUERY) {
            corivo::ax::handle_query(request);
        } else if (request.method == corivo::proto::methods::AX_PROBE_SELECTION) {
            corivo::ax::handle_probe_selection(request);
        } else if (request.method == corivo::proto::methods::AX_SUBSCRIBE) {
            corivo::ax::handle_subscribe(request);
        } else if (request.method == corivo::proto::methods::AX_UNSUBSCRIBE) {
            corivo::ax::handle_unsubscribe(request);
        } else if (request.method == corivo::proto::methods::FOREGROUND_SUBSCRIBE) {
            corivo::foreground::handle_subscribe(request);
        } else if (request.method == corivo::proto::methods::FOREGROUND_UNSUBSCRIBE) {
            corivo::foreground::handle_unsubscribe(request);
        } else if (request.method == corivo::proto::methods::FOREGROUND_CURRENT) {
            corivo::foreground::handle_current(request);
        } else if (request.method == corivo::proto::methods::OCR_RUN) {
            corivo::ocr::handle_run(request);
        } else if (request.method == corivo::proto::methods::RECORDING_START) {
            corivo::audio::handle_start(request);
        } else if (request.method == corivo::proto::methods::RECORDING_STOP) {
            corivo::audio::handle_stop(request);
        } else if (request.method == corivo::proto::methods::RECORDING_LIST_MICROPHONES) {
            corivo::audio::handle_list_microphones(request);
        } else if (request.method == corivo::proto::methods::PERMISSION_STATUS) {
            corivo::permissions::handle_status(request);
        } else {
            send_error(request.id,
                       corivo::proto::error_codes::UNSUPPORTED,
                       "unknown method: " + request.method);
        }
    } catch (const std::exception& e) {
        send_error(request.id,
                   corivo::proto::error_codes::INTERNAL,
                   std::string("handler raised: ") + e.what());
    } catch (...) {
        send_error(request.id,
                   corivo::proto::error_codes::INTERNAL,
                   "handler raised non-std exception");
    }
}

} // namespace corivo::ipc

File: packages/desktop-helpers/macos/Sources/CorivoCaptureHelper/IPC/Router.swift
//
//  Router.swift
//  Top-level dispatcher for inbound RPC. Phase 0 only handles control-plane
//  methods (`ping` / `shutdown`). Domain methods are added in their own
//  phases by extending the switch.
//

import Foundation

enum Router {
    static func dispatch(_ request: RequestMessage) async {
        do {
            switch request.method {
            // Phase 0 — control plane.
            case Methods.ping:
                try await ControlHandlers.ping(request)
            case Methods.shutdown:
                try await ControlHandlers.shutdown(request)

            // Phase 1 — screen.
            case Methods.screenCapture:
                try await ScreenHandlers.capture(request)
            case Methods.screenListDisplays:
                try await ScreenHandlers.listDisplays(request)

            // Phase 2 — accessibility.
            case Methods.axQuery:
                try await AxHandlers.query(request)
            case Methods.axProbeSelection:
                try await AxHandlers.probeSelection(request)

            // Phase 3 — async subscriptions.
            case Methods.axSubscribe:
                try await AxSubscribeHandlers.subscribe(request)
            case Methods.axUnsubscribe:
                try await AxSubscribeHandlers.unsubscribe(request)
            case Methods.foregroundSubscribe:
                try await ForegroundHandlers.subscribe(request)
            case Methods.foregroundUnsubscribe:
                try await ForegroundHandlers.unsubscribe(request)
            case Methods.foregroundCurrent:
                try await ForegroundHandlers.current(request)

            // Phase 4 — OCR.
            case Methods.ocrRun:
                try await OcrHandlers.run(request)

            // Phase 5 — meeting audio recording.
            case Methods.recordingStart:
                try await RecordingHandlers.start(request)
            case Methods.recordingStop:
                try await RecordingHandlers.stop(request)
            case Methods.recordingListMicrophones:
                try await RecordingHandlers.listMicrophones(request)

            // Phase 7 — permission status.
            case Methods.permissionStatus:
                try await PermissionHandlers.status(request)

            default:
                let err = HelperError.unsupported("unknown method: \(request.method)")
                try await EventEmitter.shared.send(
                    ResponseMessage(id: request.id, error: err.toResponseError())
                )
            }
        } catch let e as HelperError {
            do {
                try await EventEmitter.shared.send(
                    ResponseMessage(id: request.id, error: e.toResponseError())
                )
            } catch {
                Log.error("response write failed for \(request.method): \(error)")
            }
        } catch {
            do {
                let internalErr = HelperError.internal(String(describing: error))
                try await EventEmitter.shared.send(
                    ResponseMessage(id: request.id, error: internalErr.toResponseError())
                )
            } catch {
                Log.error("response write failed for \(request.method): \(error)")
            }
        }
    }
}
